crates/llms/src/chat/nsql/structured_output.rs

```diff
@@ -0,0 +1,102 @@
+/*
+Copyright 2024 The Spice.ai OSS Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+     https://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+use async_openai::{
+    error::OpenAIError,
+    types::{
+        ChatCompletionRequestMessage, ChatCompletionRequestSystemMessageArgs,
+        CreateChatCompletionRequest, CreateChatCompletionRequestArgs, CreateChatCompletionResponse,
+        ResponseFormat, ResponseFormatJsonSchema,
+    },
+};
+use schemars::{schema_for, JsonSchema};
+use serde::{Deserialize, Serialize};
+use serde_json::Value;
+
+use super::SqlGeneration;
+
+/// Implementation for [`SqlGeneration`] for [`super::Chat`] models that support [`ResponseFormat::JsonSchema`].
+pub struct StructuredOutputSqlGeneration;
+
+/// Use structured output schema to generate SQL code more reliably.
+impl SqlGeneration for StructuredOutputSqlGeneration {
+    fn create_request_for_query(
+        &self,
+        model_id: &str,
+        query: &str,
+        create_table_statements: &[String],
+    ) -> Result<CreateChatCompletionRequest, OpenAIError> {
+        let prompt = format!(
+            "```SQL\n{table_create_schemas}\n-- Using valid postgres SQL, without comments, answer the following questions for the tables provided above.\n-- {user_query}",
```

Comment:
```suggestion
            "```SQL\n{table_create_schemas}\n```\n-- Using valid postgres SQL, without comments, answer the following questions for the tables provided above.\n-- {user_query}",
```